### ISSUE TITLE
Add a logger

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#41]
We're using `console.log` in a few places (which our linting rules are about to yell at; fun fact I misconfigured our linting whoops!).

We should set up a real logger soon so we can start logging with meaningful log levels.

I'm happy to take this on since I've set it up a few times; [winston](https://www.npmjs.com/package/winston) seems to still be an oft-used tool for this.

**Comment by bickelj** [comment]
It looks like winston is still number 1 and actively maintained, pino is number 2 and actively maintained, and bunyan is number 3 and stale? Something like that. Whatever is easiest to set up, which it sounds like winston is if you are setting it up. Since I've set up neither I would lean toward pino but that is based on nothing much.

**Comment by slifty** [comment]
@bickelj if you want to run with it and you wanna run with pino that's fine by me!

If you're up for Winston, here's some existing code we can use from another OTS project: https://github.com/PermanentOrg/notification-service/blob/main/src/log.ts

**Comment by slifty** [comment]
Edit: but pino sure does look light weight and easy ;)

I defer to you completely!

**Comment by bickelj** [comment]
Attempt at finding objective reasons to use pino instead: winston has 2644KiB dependencies on `npm install` vs pino which has 2276KiB dependencies. OK, the real reason for the preference is I don't see much/any setup required for pino. I will try it and see if it "just works." If not, winston for the win.

**Comment by bickelj** [comment]
Relatively painless to get request logging working:
```
$ git diff src/app.ts
diff --git a/src/app.ts b/src/app.ts
index 5b849dc..21c7670 100644
--- a/src/app.ts
+++ b/src/app.ts
@@ -1,7 +1,10 @@
 import express from 'express';
+import PinoHttp from 'pino-http';
 import { rootRouter } from './routers';
 
 const app = express();
+const requestLogger = PinoHttp();
+app.use(requestLogger);
 app.use(express.urlencoded({ extended: true }));
 app.use(express.json());
 app.use('/', rootRouter);
```

But that is not all the logging, we would also want to replace console logs, have reasonable levels, etc.

**Comment by bickelj** [comment]
Likewise for plain logging, it was relatively painless:
```
$ git diff src/index.ts
diff --git a/src/index.ts b/src/index.ts
index ff96239..454b8bf 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -1,6 +1,10 @@
 import dotenv from 'dotenv';
+import pino from 'pino';
 import { app } from './app';
 
+const logger = pino();
+logger.info('Hello, I am logging');
+
 dotenv.config();
 
 const port = Number(process.env.PORT ?? 3000);
```

The messages are a little weird with the level being a number rather than a string, but I suppose that could be convenient at times:
`{"level":30,"time":1657661635200,"pid":54360,"hostname":"redacted","msg":"Hello, I am logging"}`

**Comment by bickelj** [comment]
What about levels, do I need to set that up? No, it works as expected, defaults to `info` level:
```$ git diff src/index.ts
diff --git a/src/index.ts b/src/index.ts
index ff96239..a10fbef 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -1,6 +1,11 @@
 import dotenv from 'dotenv';
+import pino from 'pino';
 import { app } from './app';
 
+const logger = pino();
+logger.debug('Whoa, I am debug logging');
+logger.info('Hello, I am info logging');
+logger.warn('Look out, I am warn logging!');
 dotenv.config();
```

Results in:
```
$ npm run start

> @pdc/service@0.0.0 start
> node dist/index.js

{"level":30,"time":1657661865077,"pid":54585,"hostname":"redacted","msg":"Hello, I am info logging"}
{"level":40,"time":1657661865078,"pid":54585,"hostname":"redacted","msg":"Look out, I am warn logging!"}
Server running on localhost:3000
```

**Comment by bickelj** [comment]
With `pino-pretty`, the issue is resolved. Nice.
```
$ git diff package.json
diff --git a/package.json b/package.json
index a07eed9..95c7be7 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     "migrate:dev": "ts-node -r dotenv/config src/scripts/migrate.ts",
     "test": "jest",
     "start": "node dist/index.js",
-    "start:dev": "ts-node src/index.ts"
+    "start:dev": "ts-node src/index.ts | pino-pretty"
   },
   "repository": {
     "type": "git",
@@ -49,6 +49,7 @@
     "eslint-plugin-import": "^2.26.0",
     "eslint-plugin-sort-exports": "^0.6.0",
     "jest": "^28.1.1",
+    "pino-pretty": "^8.1.0",
     "rimraf": "^3.0.2",
     "supertest": "^6.2.4",
     "ts-jest": "^28.0.5",
@@ -58,6 +59,7 @@
   "dependencies": {
     "dotenv": "^16.0.1",
     "express": "^4.18.1",
+    "pino": "^8.1.0",
     "postgres-migrations": "^5.3.0",
     "tinypg": "^7.0.0"
   }
```
$ npm run start:dev

> @pdc/service@0.0.0 start:dev
> ts-node src/index.ts | pino-pretty

[1657662062582] INFO (54903 on redacted): Hello, I am info logging
[1657662062583] WARN (54903 on redacted): Look out, I am warn logging!
Server running on localhost:3000
```

**Comment by bickelj** [comment]
What about RFC3339 formatting, though? This is a little more tricky, tried these which did not work:
```
> ts-node src/index.ts | pino-pretty --translateTime isoUtcDateTime

[i39+0000UpcDapePi7e] INFO (55139 on redacted): Hello, I am info logging
```

```
> ts-node src/index.ts | pino-pretty --translateTime UTC:yyyy-mm-dd'T'HH:MM:ss'Z'

[2022-07-12P21:46:50UTC] INFO (55316 on redacted): Hello, I am info logging
```
```
> ts-node src/index.ts | pino-pretty --translateTime UTC:yyyy-mm-ddTHH:MM:ssZ

[2022-07-12P21:48:06UTC] INFO (55442 on redacted): Hello, I am info logging
```
```
> ts-node src/index.ts | pino-pretty --translateTime UTC:yyyy-mm-dd"T"HH:MM:ss"Z"

[2022-07-12P21:49:35UTC] INFO (55586 on redacted): Hello, I am info logging
```

**Comment by bickelj** [comment]
From https://github.com/pinojs/pino/blob/v8.1.0/docs/api.md#timestamp-boolean--function
> Caution: attempting to format time in-process will significantly impact logging performance.
Ah. Perhaps that is one of the ways pino gets better performance. But the `pino-pretty` running out of process is consistent with this suggestion.

**Comment by bickelj** [comment]
This is closer, but still not quite RFC3339:
```
> ts-node src/index.ts | pino-pretty --translateTime

[2022-07-12 22:02:48.311 +0000] INFO (56282 on redacted): Hello, I am info logging
```

**Comment by bickelj** [comment]
This works when skipping the `package.json` so perhaps it is a matter of escaping characters properly:
```
$ node node_modules/ts-node/dist/bin.js src/index.ts | node node_modules/pino-pretty/bin.js --translateTime 'UTC:yyyy-MM-dd"T"HH:mm:ss.l"Z"'
[2022-07-12T22:07:38.472Z] INFO (56474 on ron): Hello, I am info logging
```

**Comment by bickelj** [comment]
And voila:
```
$ npm run start:dev

> @pdc/service@0.0.0 start:dev
> ts-node src/index.ts | pino-pretty --translateTime 'UTC:yyyy-MM-dd"T"HH:mm:ss.l"Z"'

[2022-10-12T22:07:41.900Z] INFO (56644 on redacted): Hello, I am info logging
```

```
$ git diff package.json
diff --git a/package.json b/package.json
index a07eed9..86309f6 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     "migrate:dev": "ts-node -r dotenv/config src/scripts/migrate.ts",
     "test": "jest",
     "start": "node dist/index.js",
-    "start:dev": "ts-node src/index.ts"
+    "start:dev": "ts-node src/index.ts | pino-pretty --translateTime 'UTC:yyyy-MM-dd\"T\"HH:mm:ss.l\"Z\"'"
   },
```

**Comment by bickelj** [comment]
There are still a couple of incorrect things in that format, fixing it.

**Comment by bickelj** [comment]
Better, mind your `M` and `m`:
```
$ npm run migrate:dev

> @pdc/service@0.0.0 migrate:dev
> ts-node -r dotenv/config src/scripts/migrate.ts | pino-pretty --translateTime 'UTC:yyyy-mm-dd"T"HH:MM:ss.l"Z"'

[2022-07-12T22:26:28.973Z] INFO (57854 on redacted): Starting migrations...
[2022-07-12T22:26:28.993Z] ERROR (57854 on redacted): Migrations failed!
[2022-07-12T22:26:28.993Z] ERROR (57854 on redacted): connect ECONNREFUSED 127.0.0.1:5432
```

**Comment by slifty** [comment]
Resolved by #45
